llvm/lib/Analysis/CostModel.cpp

```diff
@@ -42,25 +44,31 @@ static cl::opt<bool> TypeBasedIntrinsicCost("type-based-intrinsic-cost",
     cl::desc("Calculate intrinsics cost based only on argument types"),
     cl::init(false));
 
+static cl::opt<bool> LibCallBasedIntrinsicCost(
+    "libcall-based-intrinsic-cost",
+    cl::desc("Calculate intrinsics cost using target library info"),
+    cl::init(false));
+
 #define CM_NAME "cost-model"
 #define DEBUG_TYPE CM_NAME
 
 PreservedAnalyses CostModelPrinterPass::run(Function &F,
                                             FunctionAnalysisManager &AM) {
   auto &TTI = AM.getResult<TargetIRAnalysis>(F);
+  auto &TLI = AM.getResult<TargetLibraryAnalysis>(F);
   OS << "Printing analysis 'Cost Model Analysis' for function '" << F.getName() << "':\n";
   for (BasicBlock &B : F) {
     for (Instruction &Inst : B) {
       // TODO: Use a pass parameter instead of cl::opt CostKind to determine
       // which cost kind to print.
       InstructionCost Cost;
       auto *II = dyn_cast<IntrinsicInst>(&Inst);
-      if (II && TypeBasedIntrinsicCost) {
-        IntrinsicCostAttributes ICA(II->getIntrinsicID(), *II,
-                                    InstructionCost::getInvalid(), true);
+      if (II && (LibCallBasedIntrinsicCost || TypeBasedIntrinsicCost)) {
+        IntrinsicCostAttributes ICA(
+            II->getIntrinsicID(), *II, InstructionCost::getInvalid(),
+            /*TypeBasedOnly=*/TypeBasedIntrinsicCost, &TLI);
```

Comment:
You don't know if this is a libcall or not here, I also don't understand why this is guarded by TypeBasedIntrinsicCost. Can we remove that one too?

Reply:
I only use the `getIntrinsicInstrCost` code path if one of the flags (`type-based-intrinsic-cost` or `libcall-based-intrinsic-cost`) is set as I discovered some targets return a different cost depending on if you call `getIntrinsicInstrCost` directly or via `getInstructionCost`. This seemed to be due to them modifying the cost returned from `getIntrinsicInstrCost` within their implementation of `getInstructionCost`. I didn't want to change tests that depend on this (and it's not relevant for AArch64), so I added a flag.

Reply:
I must admit though it does seem a little odd now because `-type-based-intrinsic-cost` also implies libcall-based-intrinsic-cost since it causes `TLI` to be passed in. How about renaming `libcall-based-intrinsic-cost` to be `prefer-intrinsic-cost` or `use-intrinsic-cost`, since it's not really tied to the library calls?

This is just a suggestion, but if we had a `use-intrinsic-cost` option we could actually make it an enum along the lines of:

```
enum IntrinsicCostType {
  None,
  NonTypeBased,
  TypeBased,
};
```

that way you can collapse the two options into one and both `TypeBased` and `NonTypeBase` variants will use the TLI. However, I appreciate that would significantly increase the number of test files changed so it's possibly something for a follow-on?

Reply:
I've renamed the flag `prefer-intrinsic-cost` for this PR. The enum works nicely :+1:, but since it does increase the churn a bit I agree it's best to post that as a small follow-up PR.

Reply:
Just to be clear this flag only affects the test cost-model print pass, not LLVM more generally. The loop vectorizer in https://github.com/llvm/llvm-project/pull/128035 always passes the TLI.

Reply:
Changing the default would not be much trouble though -- it just requires adding a flag to a few tests.

Reply:
The problem I see with changing the default `prefer-intrinsic-cost` to true is that some tests may genuinely want to exercise the `getInstructionCost` path, in which case they have to live without using the TLI. I think it makes sense to follow this up in a separate PR, along with potentially changing it to use an enum instead of a boolean so that we can remove the `type-based-intrinsic-cost` flag as well?

Reply:
Yep, I'm happy to address this in the follow-up PR (since there will be a little test churn in that patch anyway).

Reply:
> Just to be clear this flag only affects the test cost-model print pass (so users = LLVM developers), not LLVM more generally. The loop vectorizer in https://github.com/llvm/llvm-project/pull/128035 always passes the TLI.

Ok that's good, thanks. I think ideally we would get rid of the option as follow-up, as it seems confusing to have the cost model printer show something that's different to the cost that is actually used in the passes making the queries.